### DISCUSSION TITLE
JSR-380 validation before serialization

### DISCUSSION THREAD

**Discussion by mluckam** [FasterXML/jackson-databind#4482] [General]
I am attempting to enforce JSR-380 validation after deserialization and before serialization without having a custom deserializer and serializer for each class.  This is achievable before deserialization as shown in [this baeldung article](https://www.baeldung.com/java-object-validation-deserialization).  Taking the same approach with serialization does not work because [BeanSerializer.serialize](https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/tools/jackson/databind/ser/BeanSerializer.java#L150) is final.

A few questions:
1.  Is there a suggested strategy to enforce validation before serialization?
2. Is there a reason that [BeanSerializer.serialize](https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/tools/jackson/databind/ser/BeanSerializer.java#L150) is final?

Below is an example project showing this with two unit tests.  Deserialization passes and serialization fails because serialization is not configured for validation.
[sample project](https://github.com/mluckam/example_project/tree/validation_during_serialization)

**Comment by cowtowncoder** [comment]
There is probably no strict reason for `BeanSerializer.serialize()` being final -- class should not be sub-classed by any application/client code, but it should be possible for proxies or such to be generated.

So I'd be ok in removing `final` from there.

As to validation, I don't think there is any good hooks to add this in: this should not require sub-classing of serializers. But there is also the problem of requiring more metadata as context that may not be easy to solve.

Overall I think currently validation is best run separate from deserialization; but I'd be open to adding additional support for in-flow extension points. But I probably won't have time driving such efforts; can and will help as usual.

**Reply by cowtowncoder** [reply]
NOTE: `BeanSerializer.serialize()` no longer final since 2.18.0.

**Comment by goekay** [comment]
hey all, i needed this today and here is my attempt: https://github.com/steve-community/ocpp-jaxb/pull/25

i was coming from the same baeldung article, and was able to create validation for serialization with a similar approach. therefore, the `final` access modifier was not a blocker for me. it works for my unit tests, but i am not sure how generalistic the impl is (i.e. whether there is an edge case it does not cover). hence, my message here. if someone finds something, please ping.

**Comment by mluckam** [comment]
@goekay 

In my experience migrating to jackson3, I found a few edge cases in de-serialization to point out:
* In `BeanDeserializerModifierWithValidation` performing `instanceof BeanDeserializer`, [your source](https://github.com/steve-community/ocpp-jaxb/pull/25/changes#diff-81000aa78c47be3bee959628b0ec9eee95a9be4f35c07460331f2dbc28cd55a3R25) misses out on deserializers that do not directly extend `BeanDeserializer`, ex `BuilderBasedDeserializer`.  Instead the check should be `instanceof BeanDeserializerBase`.  It appears you chose this strategy with serialization, [your source](https://github.com/steve-community/ocpp-jaxb/pull/25/changes#diff-79e916967e36bc576d50ac4ecdba4dd6f359155ef4fdf8836e2df46740b454fcR21).
* With the above change performed, `BeanDeserializerWithValidation` needs to extend `DelegatingDeserializer` to allow for the passing in of `ValueDeserializer` instead of `BeanDeserializer`.  Using a delegate also addresses issues that arise with `@JsonUnwrapped`.

Looking at your serialization implementation I see that jackson-databind chose to implement a [DelegatingDeserializer](https://github.com/FasterXML/jackson-databind/blob/3.x/src/main/java/tools/jackson/databind/deser/std/DelegatingDeserializer.java) but not a `DelegatingSerializer`.  There is a `StdDelegatingSerializer`, but this is not the same thing and a `StdDelegatingDeserializer` exists also.  Having a `DelegatingSerializer` would also solve some of the edge cases mentioned above.  Not sure why there was not symmetry of implementation, but possibly there is a reason.

**Reply by goekay** [reply]
@mluckam thanks for taking a look.

to be honest, i left the deserializer part as the baeldung article presented. i am aware of the delta you mention.

my initial attempt was with `instanceof BeanDeserializerBase` and my validating deserializer also extending `BeanDeserializerBase`. however, BeanDeserializerBase has so many abstract methods that need to be implemented:

<img width="881" height="284" alt="Screenshot 2026-01-31 at 18 25 49" src="https://github.com/user-attachments/assets/e1aa9d13-334a-446c-b4ae-d236ed0cd492" />

and some of them are `protected` such that i cannot delegate to delegate's same method. very annoying.

then, for symmetry, i wanted deser approach use the same approach i am using in ser approach: extending `ValueDeserializer<Object>` and wrapping the original deserializer and delegating to it. however, i was facing errors as demonstrated in [this PR's actions](https://github.com/steve-community/ocpp-jaxb/pull/26/commits).

```
Caused by: java.lang.NullPointerException: Cannot invoke "tools.jackson.core.sym.PropertyNameMatcher.matchName(String)" because "matcher" is null
	at tools.jackson.core.base.ParserMinimalBase.currentNameMatch(ParserMinimalBase.java:498)
	at tools.jackson.databind.deser.bean.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:528)
	at tools.jackson.databind.deser.bean.BeanDeserializer.deserialize(BeanDeserializer.java:200)
	at de.rwth.idsg.ocpp.jaxb.validation.BeanDeserializerWithValidation.deserialize(BeanDeserializerWithValidation.java:24)
```

at the end of the day, and after both of these attempts failing, i left the deser part as is.

it is kind of surprising how many hurdles there are just to add a simple bean validation hook :)

edit: [here is the impl](https://github.com/steve-community/ocpp-jaxb/pull/27) adapted to your observations/suggestions. i hope i understood and followed them correctly. tests fail.

**Reply by mluckam** [reply]
I think that code could help more than words in this case.  Here is an example of my implementation, [serialize-validation-injection](https://github.com/mluckam/serialize-validation-injection).  A few differences between our two implementations.

**Reply by goekay** [reply]
thanks @mluckam adjusted my impl after your example: https://github.com/steve-community/ocpp-jaxb/pull/28/changes

also added overridden impl for `deserializeWithType(JsonParser p, DeserializationContext ctxt, TypeDeserializer typeDeserializer)` for completeness.
